packages/ec2-metadata-service/src/MetadataService.ts

```diff
@@ -0,0 +1,115 @@
+        },
+        ec2MetadataV1Disabled:
+          options?.ec2MetadataV1Disabled ?? (await loadConfig(IMDSv1_DISABLED_SELECTORS, { profile })()),
+      };
+    })();
+    this.disableFetchToken = options?.disableFetchToken || false;
+  }
+
+  async request(path: string, options: { method?: string; headers?: Record<string, string> }): Promise<string> {
+    const { endpoint, ec2MetadataV1Disabled } = await this.config;
+    const handler = new NodeHttpHandler();
+    const endpointUrl = new URL(endpoint!);
+    const headers = options.headers || {};
+    /**
+     * If IMDSv1 is disabled and disableFetchToken is true, throw an error
+     * Refer: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/configuring-instance-metadata-service.html
+     */
+    if (this.disableFetchToken && ec2MetadataV1Disabled) {
+      throw new Error("IMDSv1 is disabled and fetching token is disabled, cannot make the request.");
+    }
+    /**
+     * Make request with token if disableFetchToken is not true (IMDSv2).
+     * Note that making the request call with token will result in an additional request to fetch the token.
+     */
+    if (!this.disableFetchToken) {
+      try {
+        headers["x-aws-ec2-metadata-token"] = await this.fetchMetadataToken();
+      } catch (err) {
+        if (ec2MetadataV1Disabled) {
+          // If IMDSv1 is disabled and token fetch fails (IMDSv2 fails), rethrow the error
+          throw err;
+        }
+        // If token fetch fails and IMDSv1 is not disabled, proceed without token (IMDSv1 fallback)
+      }
+    } // else, IMDSv1 fallback mode
+    const request = new HttpRequest({
+      method: options.method || "GET", // Default to GET if no method is specified
+      headers: headers,
+      hostname: endpointUrl.hostname,
+      path: endpointUrl.pathname + path,
+      protocol: endpointUrl.protocol,
+    });
+    try {
+      const { response } = await handler.handle(request, {} as HttpHandlerOptions);
+      if (response.statusCode === 200 && response.body) {
+        // handle response.body as stream
+        return sdkStreamMixin(response.body).transformToString();
+      } else {
+        throw new Error(`Request failed with status code ${response.statusCode}`);
+      }
+    } catch (error) {
+      throw new Error(`Error making request to the metadata service: ${error}`);
+    }
+  }
+
+  async fetchMetadataToken(): Promise<string> {
+    /**
+     * Refer: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instance-metadata-v2-how-it-works.html
+     */
+    const { endpoint } = await this.config;
+    const handler = new NodeHttpHandler();
+    const endpointUrl = new URL(endpoint!);
+    const tokenRequest = new HttpRequest({
+      method: "PUT",
+      headers: {
+        "x-aws-ec2-metadata-token-ttl-seconds": "21600", // 6 hours;
+      },
+      hostname: endpointUrl.hostname,
+      path: "/latest/api/token",
+      protocol: endpointUrl.protocol,
+    });
```

Comment:
nit: newlines for readability
```suggestion

    const tokenRequest = new HttpRequest({
      method: "PUT",
      headers: {
        "x-aws-ec2-metadata-token-ttl-seconds": "21600", // 6 hours;
      },
      hostname: endpointUrl.hostname,
      path: "/latest/api/token",
      protocol: endpointUrl.protocol,
    });

```